feat: choose signing cert explicitly in saml metadata ENT-677

binodpant opened the pull request:
# Why 

When the admin form is used to save / update a SAML Provider Config entity, if the option is enabled, the edx-platform code executes a task called `fetch_saml_metadata` which attempts to parse the SAML Metadata for the IDP provider indicated, and extract some relevant information. One of these pieces is the public cert, which is present in the signing key of the `KeyDescriptor` element. 

Before this change, we could accidentally choose the 'encryption' cert instead of the 'signing' cert which is the desired public key value in most cases.

This change makes that choice explicit, thus avoiding bugs and annoying customer support cycles, which wastes time for both ECS and customers.

Also, there is a case when there is no 'use' attribute, since it's optional. In that case we fall back to what we do today to capture the KeyDescriptor that contains no 'use' attribute. In this case, it looks like it is a shortcut for both signing and encryption. 

# How
When parsing, we want to explicitly choose the use='signing' cert instead of accidentally choosing the encryption one

ENT-677

# Testing notes:

Added an automated test for the `parse_metadata_xml` method, and moved it to utils

Manually tested by:

- Creating a SAML Provider Config (along with other configs) in local devstack per these instructions: https://openedx.atlassian.net/wiki/spaces/SOL/pages/131531061/How+to+set+up+Enterprise+SSO+for+Testing
- Enabling the task option so we can hit the break for parsing metadata: `Enable automatic metadata refresh`
- Hitting save and continue editing on the admin form

We now hit the parsing code where I debugged and added a attribute search for use='signing' when Choosing the `KeyDescriptor` element

<!--

🍁🍁
🍁🍁🍁🍁         🍁 Note: the Maple master branch has been created.  Please consider whether your change
    🍁🍁🍁🍁     should also be applied to Maple. If so, make another pull request against the
🍁🍁🍁🍁         open-release/maple.master branch, or ping @nedbat for help or questions.
🍁🍁

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply. You may link to information rather than copy it.
More details about the template are at https://github.com/edx/open-edx-proposals/pull/180
(link will be updated when that document merges)
-->

binodpant committed:
feat: choose signing cert explicitly in saml metadata

when parsing, we want to explicitly choose the use='signing' cert instead of accidentally choosing the encryption one

ENT-677

mattdrayer reviewed (approved):
Looks good -- we should include a test to protect against the match criteria being accidentally changed in the future

binodpant committed:
test: test for parse_metadata_xml

binodpant committed:
feat: also address case when signing key is absent

in this case we consider the 'use'-less keydescriptior as if it's for both signing and encryption

binodpant committed:
test: test case for missing use=signing attribute

mattdrayer reviewed (approved):
Looks good to me!

binodpant committed:
test: fix test failure

binodpant commented on common/djangoapps/third_party_auth/utils.py:
```diff
@@ -2,14 +2,87 @@
+    if 'urn:oasis:names:tc:SAML:2.0:protocol' not in sso_desc.get("protocolSupportEnumeration"):
+        raise MetadataParseError("This IdP does not support SAML 2.0")
+
+    # Now we just need to get the public_key and sso_url
+    # We want the use='signing' cert, not the 'encryption' one
+    public_key = sso_desc.findtext("./{}[@use='signing']//{}".format(
+        etree.QName(SAML_XML_NS, "KeyDescriptor"), "{http://www.w3.org/2000/09/xmldsig#}X509Certificate"
+    ))
+    if not public_key:
+        # it's possible that there is just one keyDescription with no use attribute
+        # that is a shortcut for both signing and encryption combined. So we can use that as fallback.
+        public_key = sso_desc.findtext("./{}//{}".format(
+            etree.QName(SAML_XML_NS, "KeyDescriptor"), "{http://www.w3.org/2000/09/xmldsig#}X509Certificate"
+        ))
+        if not public_key:
+            raise MetadataParseError("Public Key missing. Expected an <X509Certificate>")
```
this here, is the essence of this change, just mentioning since this method was moved out from tasks.py to here.. so changes may not have been apparent

alex-sheehan-edx reviewed (approved):
lgtm!
